### ISSUE TITLE
cassandra-reaper does not launch on docker with cassandra backend

### ISSUE THREAD

**Issue by sguenfred** [thelastpickle/cassandra-reaper#830]
Hi, it seems that the latest docker packaged version does not perform schema update when starting from a fresh db

i just created the db and it s working perfectly
```
ssh 10.46.1.72 nodetool -u cassandra -pw cassandra status reaptest
Datacenter: north
=================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address     Load       Tokens       Owns (effective)  Host ID                               Rack
UN  10.46.1.88  7.28 GiB   256          100.0%            b29908a2-b28a-442b-90d7-06921b215595  west
UN  10.46.1.72  7.29 GiB   256          100.0%            20e3336a-1484-40bf-8931-e78b80610e64  east
Datacenter: south
=================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address     Load       Tokens       Owns (effective)  Host ID                               Rack
UN  10.46.1.58  8.73 GiB   256          100.0%            33644628-3c10-4777-a495-fdae4722cfa4  west
UN  10.46.1.97  2.11 GiB   256          100.0%            3afc1744-dad7-4508-8e4d-deb33dec4975  east
```

but when i launch the docker

```docker run -p 8080:8080     -p 8081:8081     -e "REAPER_JMX_AUTH_USERNAME=cassandra"     -e "REAPER_JMX_AUTH_PASSWORD=cassandra"     -e "REAPER_STORAGE_TYPE=cassandra"     -e "REAPER_CASS_CLUSTER_NAME=reaptest"     -e "REAPER_CASS_CONTACT_POINTS=[\"10.46.1.72\"]"     -e "REAPER_CASS_KEYSPACE=reaptest"     thelastpickle/cassandra-reaper:latest```

i got 

```
+ touch /etc/cassandra-reaper.yml
+ /usr/local/bin/configure-persistence.sh
+ /usr/local/bin/configure-webui-authentication.sh
+ /usr/local/bin/configure-metrics.sh
+ exec java -jar /usr/local/lib/cassandra-reaper.jar server /etc/cassandra-reaper.yml
INFO   [2020-01-09 14:03:46,867] [main] i.d.s.DefaultServerFactory - Registering jersey handler with root path prefix: / 
INFO   [2020-01-09 14:03:46,870] [main] i.d.s.DefaultServerFactory - Registering admin handler with root path prefix: / 
INFO   [2020-01-09 14:03:46,870] [main] i.d.a.AssetsBundle - Registering AssetBundle with name: assets for path /webui/* 
INFO   [2020-01-09 14:03:46,948] [main] o.a.s.c.ReflectionBuilder - An instance with name 'authc' already exists.  Redefining this object as a new instance of type org.apache.shiro.web.filter.authc.PassThruAuthenticationFilter 
INFO   [2020-01-09 14:03:46,949] [main] o.a.s.c.ReflectionBuilder - An instance with name 'rest' already exists.  Redefining this object as a new instance of type io.cassandrareaper.resources.auth.RestPermissionsFilter 
INFO   [2020-01-09 14:03:46,992] [main] i.c.ReaperApplication - initializing runner thread pool with 15 threads 
INFO   [2020-01-09 14:03:46,992] [main] i.c.ReaperApplication - initializing storage of type: cassandra 
INFO   [2020-01-09 14:03:47,018] [main] c.d.d.core - DataStax Java driver 3.7.1 for Apache Cassandra 
INFO   [2020-01-09 14:03:47,020] [main] c.d.d.c.GuavaCompatibility - Detected Guava >= 19 in the classpath, using modern compatibility layer 
INFO   [2020-01-09 14:03:47,172] [main] c.d.d.c.ClockFactory - Using native clock to generate timestamps. 
INFO   [2020-01-09 14:03:47,243] [main] c.d.d.c.NettyUtil - Did not find Netty's native epoll transport in the classpath, defaulting to NIO. 
INFO   [2020-01-09 14:03:47,902] [main] c.d.d.c.p.DCAwareRoundRobinPolicy - Using data-center name 'north' for DCAwareRoundRobinPolicy (if this is incorrect, please provide the correct datacenter name with DCAwareRoundRobinPolicy constructor) 
INFO   [2020-01-09 14:03:47,903] [main] c.d.d.c.Cluster - New Cassandra host /10.46.1.58:9042 added 
INFO   [2020-01-09 14:03:47,904] [main] c.d.d.c.Cluster - New Cassandra host /10.46.1.72:9042 added 
INFO   [2020-01-09 14:03:47,904] [main] c.d.d.c.Cluster - New Cassandra host /10.46.1.88:9042 added 
INFO   [2020-01-09 14:03:47,904] [main] c.d.d.c.Cluster - New Cassandra host /10.46.1.97:9042 added 
INFO   [2020-01-09 14:03:48,074] [main] o.c.c.m.MigrationRepository - Found 8 migration scripts 
WARN   [2020-01-09 14:03:48,074] [main] i.c.s.CassandraStorage - Starting db migration from 16 to 23… 
ERROR  [2020-01-09 14:03:48,081] [main] i.c.ReaperApplication - Storage is not ready yet, trying again to connect shortly... 
com.datastax.driver.core.exceptions.InvalidQueryException: unconfigured table running_reapers
	at com.datastax.driver.core.exceptions.InvalidQueryException.copy(InvalidQueryException.java:49)
	at com.datastax.driver.core.DriverThrowables.propagateCause(DriverThrowables.java:35)
	at com.datastax.driver.core.DefaultResultSetFuture.getUninterruptibly(DefaultResultSetFuture.java:293)
	at com.datastax.driver.core.AbstractSession.execute(AbstractSession.java:58)
	at com.datastax.driver.core.AbstractSession.execute(AbstractSession.java:39)
	at io.cassandrareaper.storage.CassandraStorage.initializeAndUpgradeSchema(CassandraStorage.java:254)
	at io.cassandrareaper.storage.CassandraStorage.<init>(CassandraStorage.java:227)
	at io.cassandrareaper.ReaperApplication.initializeStorage(ReaperApplication.java:439)
	at io.cassandrareaper.ReaperApplication.tryInitializeStorage(ReaperApplication.java:310)
	at io.cassandrareaper.ReaperApplication.run(ReaperApplication.java:172)
	at io.cassandrareaper.ReaperApplication.run(ReaperApplication.java:91)
	at io.dropwizard.cli.EnvironmentCommand.run(EnvironmentCommand.java:43)
	at io.dropwizard.cli.ConfiguredCommand.run(ConfiguredCommand.java:87)
	at io.dropwizard.cli.Cli.run(Cli.java:78)
	at io.dropwizard.Application.run(Application.java:93)
	at io.cassandrareaper.ReaperApplication.main(ReaperApplication.java:110)
Caused by: com.datastax.driver.core.exceptions.InvalidQueryException: unconfigured table running_reapers
	at com.datastax.driver.core.Responses$Error.asException(Responses.java:181)
	at com.datastax.driver.core.DefaultResultSetFuture.onSet(DefaultResultSetFuture.java:215)
	at com.datastax.driver.core.RequestHandler.setFinalResult(RequestHandler.java:228)
	at com.datastax.driver.core.RequestHandler.access$2600(RequestHandler.java:62)
	at com.datastax.driver.core.RequestHandler$SpeculativeExecution.setFinalResult(RequestHandler.java:1005)
	at com.datastax.driver.core.RequestHandler$SpeculativeExecution.onSet(RequestHandler.java:808)
	at com.datastax.driver.core.Connection$Dispatcher.channelRead0(Connection.java:1233)
	at com.datastax.driver.core.Connection$Dispatcher.channelRead0(Connection.java:1151)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:356)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:342)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:335)
	at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:356)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:342)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:335)
	at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:356)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:342)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:335)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:312)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:286)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:356)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:342)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:335)
	at io.netty.channel.ChannelInboundHandlerAdapter.channelRead(ChannelInboundHandlerAdapter.java:86)
	at com.datastax.driver.core.InboundTrafficMeter.channelRead(InboundTrafficMeter.java:38)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:356)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:342)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:335)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1304)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:356)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:342)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:921)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:135)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:646)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:581)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:498)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:460)
	at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:131)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)```

it seems that cassandra reaper is accessing the keyspace but he is unable to perform schema update, any clues?

**Comment by adejanovski** [comment]
Hi, 

could you please provide the output of the following queries?

- `desc keyspace reaptest`
- `select applied_successful, version, executed_at from reaptest.schema_migration`

Thanks

**Comment by sguenfred** [comment]
Sure, here is the keyspace description
```
cassandra@cqlsh> desc KEYSPACE reaptest;

CREATE KEYSPACE reaptest WITH replication = {'class': 'NetworkTopologyStrategy', 'north': '2', 'south': '2'}  AND durable_writes = true;

CREATE TABLE reaptest.schema_migration (
    applied_successful boolean,
    version int,
    executed_at timestamp,
    script text,
    script_name text,
    PRIMARY KEY (applied_successful, version)
) WITH CLUSTERING ORDER BY (version ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy', 'max_threshold': '32', 'min_threshold': '4'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';

CREATE TABLE reaptest.schema_migration_leader (
    keyspace_name text PRIMARY KEY,
    leader uuid,
    leader_hostname text,
    took_lead_at timestamp
) WITH bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy', 'max_threshold': '32', 'min_threshold': '4'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';
```
and the result the migration history

```
cassandra@cqlsh> select applied_successful, version, executed_at from reaptest.schema_migration
   ... ;

 applied_successful | version | executed_at
--------------------+---------+---------------------------------
              False |      17 | 2020-01-09 11:29:39.859000+0000
               True |      16 | 2020-01-09 11:29:39.784000+0000

(2 rows)
```

thanks for helping

**Comment by adejanovski** [comment]
Did you start several Reaper instances at the same time or just one?

There's something very weird here which is that you only have the base migration tables in there but none of the Reaper tables that should've been created by migration 16, which is reported as applied.

Could you run `nodetool describecluster` and see if all nodes are on the same schema version?
We may have a split brain situation due to a faulty leader election in the migration lib, that lets concurrent Reapers in different DCs migrate the schema concurrently.
Also check that the schema for the reaptest keyspace is the same on all nodes.

If you're in none of the above cases, truncate the schema_migration table and restart Reaper to apply all the migrations.

**Comment by sguenfred** [comment]
This cassandra cluster already have a old cassandra reaper operating on it, but of course the other cassandra reaper have his config on another keyspace.  
This is just a teporary situation for the migration and of course i just want to see the new ui and i will not schedule any repair. But maybe having two cassandra reaper backend on the same cassandra cluster is not supported?

anyway here is the nodetools result
```
nodetool -u cassandra -pw cassandra describecluster
Cluster Information:
	Name: cassandra-test
	Snitch: org.apache.cassandra.locator.GossipingPropertyFileSnitch
	DynamicEndPointSnitch: enabled
	Partitioner: org.apache.cassandra.dht.Murmur3Partitioner
	Schema versions:
		689a8f46-85d1-316c-8d97-8561e5a89649: [10.46.1.58, 10.46.1.88, 10.46.1.72, 10.46.1.97]
```
i gonna try the truncate schema_migration solution and let you know

**Comment by adejanovski** [comment]
You can have several Reaper clusters using the same cluster as backend, on different keyspaces. That should not be what's causing your issue
Let me know how it goes once you truncated the `reaptest.schema_migration` table.

**Comment by sguenfred** [comment]
just done it
```
cassandra@cqlsh:reaptest> truncate reaptest.schema_migration;
cassandra@cqlsh:reaptest> select applied_successful, version, executed_at from reaptest.schema_migration ;

 applied_successful | version | executed_at
--------------------+---------+-------------

(0 rows)
```
then lauch cassandra-reaper
```
docker run -p 8080:8080     -p 8081:8081     -e "REAPER_JMX_AUTH_USERNAME=cassandra"     -e "REAPER_JMX_AUTH_PASSWORD=cassandra"     -e "REAPER_STORAGE_TYPE=cassandra"     -e "REAPER_CASS_CLUSTER_NAME=reaptest"     -e "REAPER_CASS_CONTACT_POINTS=[\"10.46.1.72\"]"     -e "REAPER_CASS_KEYSPACE=reaptest"     thelastpickle/cassandra-reaper:latest
```
then checked the migration status
```
cassandra@cqlsh:reaptest> select applied_successful, version, executed_at from reaptest.schema_migration ;

 applied_successful | version | executed_at
--------------------+---------+---------------------------------
              False |      17 | 2020-01-09 16:31:36.567000+0000
               True |      16 | 2020-01-09 16:31:36.486000+0000

(2 rows)
```

**Comment by adejanovski** [comment]
ok, I found the problem and it's super lame... We kept the keyspace name in the initial migration (number 16) 😅 
I'll push a new version to fix this.

**Comment by sguenfred** [comment]
Good   news :)
will a image be available on the docker hub?

**Comment by adejanovski** [comment]
@sguenfred, I just released 2.0.2 with the patch, and the `latest` image was updated in Docker hub.
You can pull it again to refresh it and give it a new try.

**Comment by sguenfred** [comment]
just tried it from a fresh install.
Migration error was gone, everything seems to be OK thanks,
i close this one
